src/multirw.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use lmdb;
+
+use std::marker::PhantomData;
+use std::rc::Rc;
+use lmdb::{Cursor, Database, Iter as LmdbIter, IterDup as LmdbIterDup, RoCursor, RoTransaction, RwTransaction,
+           Transaction};
+
+use lmdb::WriteFlags;
+
+use error::StoreError;
+
+use value::Value;
+
+fn read_transform(val: Result<&[u8], lmdb::Error>) -> Result<Option<Value>, StoreError> {
+    match val {
+        Ok(bytes) => {
+            Value::from_tagged_slice(bytes).map(Some).map_err(
+                StoreError::DataError,
+            )
+        },
+        Err(lmdb::Error::NotFound) => Ok(None),
+        Err(e) => Err(StoreError::LmdbError(e)),
+    }
+}
+
+pub struct MultiWriter<'env, K>
+where
+    K: AsRef<[u8]>,
+{
+    tx: RwTransaction<'env>,
+    phantom: PhantomData<K>,
+}
+
+pub struct MultiReader<'env, K>
+where
+    K: AsRef<[u8]>,
+{
+    tx: RoTransaction<'env>,
+    phantom: PhantomData<K>,
+}
+
+pub struct MultiIter<'env> {
+    iter: LmdbIterDup<'env>,
+    cursor: RoCursor<'env>,
+}
+
+pub struct Iter<'env> {
+    iter: LmdbIter<'env>,
+}
+
+pub struct MultiCursor<'env, K>
+where
+    K: AsRef<[u8]>,
+{
+    tx: RwTransaction<'env>,
+    phantom: PhantomData<K>,
+}
+
+impl<'env, K> MultiCursor<'env, K>
+where
+    K: AsRef<[u8]>,
+{
+    /// Provides a cursor to all of the values for the duplicate entries that match this key
+    pub fn get(&self, store: MultiStore, k: K) -> Result<Iter, StoreError> {
+        let mut cursor = self.tx.open_ro_cursor(store.0).map_err(
+            StoreError::LmdbError,
+        )?;
+        let iter = cursor.iter_dup_of(k);
+        //Ok(Iter{ iter, cursor })
+        Ok(Iter { iter })
+    }
+
+    /// Consume this MultiCursor and give the `MultiWriter` back
+    /// So that it may perform additional tasks
+    pub fn into_writer(self) -> MultiWriter<'env, K> {
+        MultiWriter {
+            tx: self.tx,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'env, K> MultiWriter<'env, K>
+where
+    K: AsRef<[u8]>,
+{
+    pub(crate) fn new(txn: RwTransaction) -> MultiWriter<K> {
+        MultiWriter {
+            tx: txn,
+            phantom: PhantomData,
+        }
+    }
+
+    /// This cursor consumes the writer, as it is not safe to attempt to access records while using
+    /// put or delete, as the location of the records might change
```

Comment:
It's possible to write records using a cursor in LMDB, although rkv doesn't yet support this. Perhaps it would be more ergonomic to support that functionality than to make the consumer convert back and forth between the MultiCursor and the MultiWriter. Then we could keep `MultiWriter.into_cursor()`, which consumes the MultiWriter, but expose `MultiCursor.put()/delete()` methods, so it isn't necessary to convert back to MultiWriter via `MultiCursor.into_writer()` in order to make changes.

Reply:
I agree, I still think we should test with the cursor to ensure that it's safe to get->put dup values in LMDB while ensuring that the retrieved value is still in-tact.  I did a significant amount of work with LMDB about 5 years ago, and I vaguely recall this exact problem, but maybe the API has been updated since then. 

Adding Cursor support would dovetail in very nicely with my other proposed API change: https://github.com/mozilla/rkv/issues/95